File: 현지/2026_05_28/순위 검색.py
from bisect import bisect_left

def solution(info, query):

    answer = []

    dic = {}

    # info 돌기
    for i in range(len(info)):

        temp = info[i].split()

        lang = temp[0]
        job = temp[1]
        career = temp[2]
        food = temp[3]
        score = int(temp[4])

        arr1 = [lang, '-']
        arr2 = [job, '-']
        arr3 = [career, '-']
        arr4 = [food, '-']

        # 16개 경우 만들기
        for a in arr1:
            for b in arr2:
                for c in arr3:
                    for d in arr4:

                        key = a + b + c + d

                        if key not in dic:
                            dic[key] = []

                        dic[key].append(score)

    # 정렬
    for k in dic:
        dic[k].sort()

    # query 처리
    for q in query:

        q = q.replace(" and ", " ")

        temp2 = q.split()

        q1 = temp2[0]
        q2 = temp2[1]
        q3 = temp2[2]
        q4 = temp2[3]
        qscore = int(temp2[4])

        find = q1 + q2 + q3 + q4

        # 없는 경우
        if find not in dic:

            answer.append(0)

        else:

            score_list = dic[find]

            # 이분탐색
            idx = bisect_left(score_list, qscore)

            cnt = len(score_list) - idx

            answer.append(cnt)

    return answer

File: 준영/2026_05_28/순위검색.java
import java.util.*;
class Solution {
    List<Node> p = new ArrayList<>();
    List<Node> p2 = new ArrayList<>();
    public class Node{
        String lang;
        String major;
        String year;
        String food;
        int score;
        public Node(String lang, String major, String year, String food, int score){
            this.lang = lang;
            this.major = major;
            this.year = year;
            this.food = food;
            this.score = score;
        }
    }

    public int[] solution(String[] info, String[] query) {

        for (int i = 0; i < info.length; i++){
            String[] new_info = info[i].split(" ");
            p.add(new Node(new_info[0], new_info[1], new_info[2], new_info[3], Integer.parseInt(new_info[4])));   // 노드에 값을 저장
        }
        for (int i = 0; i < query.length; i++){
            String[] new_query = query[i].split(" and ");
            String[] s = new_query[3].split(" ");
            p2.add(new Node(new_query[0], new_query[1], new_query[2], s[0], Integer.parseInt(s[1])));
        }
        int[] answer = new int[query.length];  // 회사의 길이만큼
        for (int i = 0; i < p2.size(); i++){
            for (int j = 0; j < p.size(); j++){
                if (!(p2.get(i).lang.equals(p.get(j).lang)) && !(p2.get(i).lang.equals("-"))){
                    // 언어가 동일하지 않으면 패스
                    continue;
                }
                else if (!(p2.get(i).major.equals(p.get(j).major)) && !(p2.get(i).major.equals("-"))){
                        continue;
                }
                else if (!(p2.get(i).year.equals(p.get(j).year)) && !(p2.get(i).year.equals("-"))){
                   continue;
                }
                else if (!(p2.get(i).food.equals(p.get(j).food)) && !(p2.get(i).food.equals("-"))){
                    continue;
                }
                else if (p2.get(i).score > p.get(j).score){
                    continue;
                }
                answer[i]++;
            }
        }
        return answer;
    }
}
